### PR TITLE
New: Enhanced pinch to zoom

### PR DESCRIPTION
**Pull request by jeremypress**
Still a WIP, but this PR adds a more fluid pinch-to-zoom by:
1. Scaling the visible pages canvases with CSS transform
2. Calculating the overall scale and translate it to a PDF.js zoom
3. Removing the CSS scale and scrolling to where we left off while pinching

Open Issues:
1. Slight jump when transitioning from CSS scaling to PDF.js zoom
2. Pages will occasionally scroll when pinching using the presentation viewer.

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @jeremypress has signed the CLA. Thanks for the pull request!

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 1]
```diff
@@ -977,6 +970,135 @@ class DocBaseViewer extends BaseViewer {
...
+    }
+
+    /**
+     * Replaces the CSS transform with a native PDF.js zoom and scrolls to maintain positioning.
+     *
+     * @protected
+     * @return {void}
+     */
+    pinchToZoomEndHandler() {
+        if (this.zoomWrapper && this.isPinching) {
```
Comment (JustinHoldstock) [inline src/lib/viewers/doc/DocBaseViewer.js]: To remove a level of arrowness, you could do `if(!this.zoomWrapper || !this.isPinching) return`

File: src/lib/util.js [hunk 2]
```diff
@@ -751,3 +751,61 @@ export function pageNumberFromScroll(currentPageNum, previousScrollTop, currentP
...
+            page.offsetLeft + page.scrollWidth,
+            page.offsetTop + page.scrollHeight
+        );
+        const distance = getDistance(pageMidpoint[0], pageMidpoint[1], x, y);
+        if (!closestPage || distance < closestPage[1]) {
+            closestPage = [page, distance];
+        }
+    }
+
+    return closestPage[0];
```
Comment (JustinHoldstock) [inline src/lib/util.js]: Careful, this'll error out if `closestPage` is never set

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 3]
```diff
@@ -977,6 +968,131 @@ class DocBaseViewer extends BaseViewer {
...
+        const yOrigin = this.docEl.scrollTop - this.pinchPage.offsetTop + touchMidpoint[1] + 15;
+
+        this.pinchPage.style['transform-origin'] = `${xOrigin}px ${yOrigin}px`;
+
+        // Calculate current scroll position and gesture offset
+        this.yOffset = yOrigin;
+        this.xOffset = xOrigin;
+        this.y = touchMidpoint[1];
+        this.x = touchMidpoint[0];
+        //
```
Comment (JustinHoldstock) [inline src/lib/viewers/doc/DocBaseViewer.js]: empty comment

File: src/lib/viewers/doc/DocBaseViewer.js [hunk 4]
```diff
@@ -977,6 +968,131 @@ class DocBaseViewer extends BaseViewer {
...
+     */
+    pinchToZoomEndHandler() {
+        if (!this.pinchPage || !this.isPinching) {
+            return;
+        }
+
+        // PDF.js zoom
+        this.pdfViewer.currentScaleValue = this.pdfViewer.currentScale * this.pinchScale;
+
+        // // Scroll to correct position after zoom
```
Comment (JustinHoldstock) [inline src/lib/viewers/doc/DocBaseViewer.js]: extra `//` in comment (unless github is acting up)
 as well as commented code

File: src/lib/util.js (hunk 2 above)
Comment (jeremypress) [reply src/lib/util.js]: check line 795!

**Pull request merged**